File: 70.爬楼梯/climbStairs.go
package problem0070

// 动态规划
// 状态转移方程 F(n) = F(n-1) + F(n-2)
// 边界        F(1), F(2)
// 最优子结构   F(8)和F(9)是F(10)的最优子结构
func climbStairs(n int) int {
	if n <= 2 {
		return n
	}
	pre1, pre2 := 1, 2
	for i := 3; i <= n; i++ {
		pre1, pre2 = pre2, pre1+pre2
	}
	return pre2
}

File: 70.爬楼梯/climbStairs.js
// 状态转移方程 = dp[n] = dp[n-1] + dp[n-2]
// 从状态转移方程可以发现，dp[n]至于前2个数值有关，因此此处可以使用滚动数组将空间复杂度优化为O(1)
function climbStairs(n) {
  if (n < 2) {
    return n
  }
  let a = 1
  let b = 2
  for (let i = 3; i <= n; i++) {
    [a, b] = [b, a+b]
  }
  return b
}
